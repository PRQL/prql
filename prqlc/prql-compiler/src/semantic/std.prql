# The PRQL standard library defines the following functions and transforms.
# The definitions are whitespace insensitive, and have this form:
#
# ```
# let my_func = param1 param2 ...  -> <return_type> body_expr
# ```
#
# Where:
#   * `my_func` is the name of the function
#   * `param1` is the first parameter optionally followed by a type in "< ... >"
#   * `param2` etc. follow the same pattern as param1
#   * `<return_type>` is the type of result wrapped in "< ... >"
#   * `body_expr` defines the function body that creates the result.
#     It can be PRQL code or `internal ...` to indicate internal compiler code.

# Operators

let mul = left right -> <int || float> internal std.mul
let div_i = left right -> <int || float> internal std.div_i
let div_f = left right -> <int || float> internal std.div_f
let mod = left right -> <int || float> internal std.mod
let add = left right -> <int || float || timestamp || date> internal std.add
let sub = left right -> <int || float || timestamp || date> internal std.sub
let eq = left right -> <bool> internal std.eq
let ne = left right -> <bool> internal std.ne
let gt = left right -> <bool> internal std.gt
let lt = left right -> <bool> internal std.lt
let gte = left right -> <bool> internal std.gte
let lte = left right -> <bool> internal std.lte
let and = left<bool> right<bool> -> <bool> internal std.and
let or = left<bool> right<bool> -> <bool> internal std.or
let coalesce = left right -> internal std.coalesce
let regex_search = text pattern -> <bool> internal std.regex_search

let neg = expr<int || float> -> <int || float> internal std.neg
let not = expr<bool> -> <bool> internal std.not

# Types

## Type primitives
type int = int
type float = float
type bool = bool
type text = text
type date = date
type time = time
type timestamp = timestamp
type `func` = func
type anytype = anytype

## Generic array
# TODO: an array of anything, not just nulls
type array = [anytype]

## Scalar
type scalar = int || float || bool || text || date || time || timestamp || null
type tuple = {anytype..}

## Range
type range = {start = scalar, end = scalar}

## Relation (an array of tuples)
type relation = [tuple]

## Transform
type transform = func relation -> relation

# Functions

## Relational transforms
let from = func
  `default_db.source` <relation>
  -> <relation> internal from

let select = func
  columns <scalar || tuple>
  tbl <relation>
  -> <relation> internal select

let filter = func
  condition <bool>
  tbl <relation>
  -> <relation> internal filter

let derive = func
  columns <scalar || tuple>
  tbl <relation>
  -> <relation> internal derive

let aggregate = func
  columns <scalar || tuple>
  tbl <relation>
  -> <relation> internal aggregate

let sort = func
  by <scalar || tuple>
  tbl <relation>
  -> <relation> internal sort

let take = func
  expr <anytype>
  tbl <relation>
  -> <relation> internal take

let join = func
  `default_db.with` <relation>
  condition <bool>
  `noresolve.side`:inner
  tbl <relation>
  -> <relation> internal join

let group = func
  by<scalar || tuple>
  pipeline <transform>
  tbl <relation>
  -> <relation> internal group

let window = func
  rows:0..0
  range:0..0
  expanding <bool>:false
  rolling <int>:0
  pipeline <transform>
  tbl <relation>
  -> <relation> internal window

let append = `default_db.bottom`<relation> top<relation> -> <relation> internal append
let intersect = `default_db.bottom`<relation> top<relation> -> <relation> (
  t = top
  join (b = bottom) (tuple_every (tuple_map _eq (tuple_zip t.* b.*)))
  select t.*
)
let remove = `default_db.bottom`<relation> top<relation> -> <relation> (
  t = top
  join side:left (b = bottom) (tuple_every (tuple_map _eq (tuple_zip t.* b.*)))
  filter (tuple_every (tuple_map _is_null b.*))
  select t.*
)
let loop = func
  pipeline <transform>
  top <relation>
  -> <relation> internal loop

## Aggregate functions
# These return either a scalar when used within `aggregate`, or a column when used anywhere else.

let min = column <array> -> <int || float || null> internal std.min

let max = column <array> -> <int || float || null> internal std.max

let sum = column <array> -> <int || float> internal std.sum

let average = column <array> -> <float || null> internal std.average

let stddev = column <array> -> <float || null> internal std.stddev

let all = column <array> -> <bool> internal std.all

let any = column <array> -> <bool> internal std.any

let concat_array = column <array> -> <text> internal std.concat_array

# Counts number of items in the column.
# Note that the count will include null values.
let count = column<array> -> <int> internal count

# Deprecated in favour of filterning input to the [std.count] function (not yet implemented).
@{deprecated}
let count_distinct = column <array> -> internal std.count_distinct

## Window functions
let lag =   offset <int>    column <array> -> internal std.lag
let lead =  offset <int>    column <array> -> internal std.lead
let first      = column <array> -> internal std.first
let last       = column <array> -> internal std.last
let rank       = column <array> -> internal std.rank
let rank_dense = column <array> -> internal std.rank_dense
let row_number = column <array> -> internal row_number

# Mathematical functions
module math {
  let abs = column -> <int || float> internal std.math.abs
  let floor = column -> <int> internal std.math.floor
  let ceil = column -> <int> internal std.math.ceil
  let pi = -> <float> internal std.math.pi
  let exp = column -> <int || float> internal std.math.exp
  let ln = column -> <int || float> internal std.math.ln
  let log10 = column -> <int || float> internal std.math.log10
  let log = base<int || float> column -> <int || float> internal std.math.log
  let sqrt = column -> <int || float> internal std.math.sqrt
  let degrees = column -> <int || float> internal std.math.degrees
  let radians = column -> <int || float> internal std.math.radians
  let cos = column -> <int || float> internal std.math.cos
  let acos = column -> <int || float> internal std.math.acos
  let sin = column -> <int || float> internal std.math.sin
  let asin = column -> <int || float> internal std.math.asin
  let tan = column -> <int || float> internal std.math.tan
  let atan = column -> <int || float> internal std.math.atan
  let pow = exponent<int || float> column -> <int || float> internal std.math.pow
  let round = n_digits column -> <scalar> internal std.math.round
}

## Misc functions
let as = `noresolve.type` column -> <scalar> internal std.as
let in = pattern value -> <bool> internal in

## Tuple functions
let tuple_every = func list -> <bool> internal tuple_every
let tuple_map = func fn <func> list -> internal tuple_map
let tuple_zip = func a b -> internal tuple_zip
let _eq = func a -> internal _eq
let _is_null = func a -> _param.a == null

## Misc
let from_text = input<text> `noresolve.format`:csv -> <relation> internal from_text

## Text functions
module text {
  let lower = column -> <text> internal std.text.lower
  let upper = column -> <text> internal std.text.upper
  let ltrim = column -> <text> internal std.text.ltrim
  let rtrim = column -> <text> internal std.text.rtrim
  let trim = column -> <text> internal std.text.trim
  let length = column -> <int> internal std.text.length
  let extract = offset<int> length<int> column -> <text> internal std.text.extract
  let replace = pattern<text> replacement<text> column -> <text> internal std.text.replace
  let starts_with = prefix<text> column -> <bool> internal std.text.starts_with
  let contains = substr<text> column -> <bool> internal std.text.contains
  let ends_with = suffix<text> column -> <bool> internal std.text.ends_with
}

## Date functions
module date {
  let to_text = format<text> column -> <text> internal std.date.to_text
}

## File-reading functions, primarily for DuckDB
let read_parquet = source<text> -> <relation> internal std.read_parquet
let read_csv = source<text> -> <relation> internal std.read_csv


## PRQL compiler functions

let prql_version = -> <text> internal prql_version
