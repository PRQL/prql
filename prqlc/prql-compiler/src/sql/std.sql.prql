#! Implementation of `std` module.
#!
#! This file is not really PRQL.
#! It can contain only:
#! - functions declarations that don't have named params and s-string-only body,
#! - module declarations whose names correspond to sql dialect names.
#!
#! Functions can define `binding_strength` annotation, which signifies how much
#! precedence does the top-level operation in the s-string provide.
#! This value defaults to 100 (high precedence).
#!
#! S-strings can define required binding strength of the interpolated expression.
#! This value defaults to binding strength of the function.



# Aggregation functions
@{window_frame=true}
let min = column -> s"MIN({column:0})"

@{window_frame=true}
let max = column -> s"MAX({column:0})"

@{window_frame=true, coalesce="0"}
let sum = column -> s"SUM({column:0})"

@{window_frame=true}
let average = column -> s"AVG({column:0})"

@{window_frame=true}
let stddev = column -> s"STDDEV({column:0})"

@{window_frame=true, coalesce="TRUE"}
let all = column -> s"BOOL_AND({column:0})"

@{window_frame=true, coalesce="FALSE"}
let any = column -> s"BOOL_OR({column:0})"

@{window_frame=true, coalesce="''"}
let concat_array = column -> s"STRING_AGG({column:0}, '')"

@{window_frame=true}
let count = column -> s"COUNT(*)"

@{window_frame=true}
let count_distinct = column -> s"COUNT(DISTINCT {column:0})"

# Window functions
let lag = offset column -> s"LAG({column:0}, {offset:0})"

let lead = offset column -> s"LEAD({column:0}, {offset:0})"

let first = column -> s"FIRST_VALUE({column:0})"

let last = column -> s"LAST_VALUE({column:0})"

let rank = -> s"RANK()"

let rank_dense = -> s"DENSE_RANK()"

let row_number = -> s"ROW_NUMBER()"

# Mathematical functions
module math {
  # Clickhouse: https://clickhouse.com/docs/en/sql-reference/functions/math-functions
  # DuckDB: https://duckdb.org/docs/test/functions/math.html
  # MySQL: https://dev.mysql.com/doc/refman/8.0/en/mathematical-functions.html
  # Postgres: https://www.postgresql.org/docs/7.4/functions-math.html
  # SQLite: https://www.sqlite.org/lang_mathfunc.html
  # MSSQL: https://learn.microsoft.com/en-us/sql/t-sql/functions/mathematical-functions-transact-sql
  # BigQuery: https://cloud.google.com/bigquery/docs/reference/standard-sql/mathematical_functions
  # Snowflake: https://docs.snowflake.com/en/sql-reference/functions-math.html
  let abs = column -> s"ABS({column:0})"
  let floor = column -> s"FLOOR({column:0})"
  let ceil = column -> s"CEIL({column:0})"
  let pi = -> s"PI()"
  let exp = column -> s"EXP({column:0})"
  let ln = column -> s"LN({column:0})"
  let log10 = column -> s"LOG10({column:0})"
  let log = base column -> s"LOG10({column:0}) / LOG10({base:0})"
  let sqrt = column -> s"SQRT({column:0})"
  let degrees = column -> s"DEGREES({column:0})"
  let radians = column -> s"RADIANS({column:0})"
  let cos = column -> s"COS({column:0})"
  let acos = column -> s"ACOS({column:0})"
  let sin = column -> s"SIN({column:0})"
  let asin = column -> s"ASIN({column:0})"
  let tan = column -> s"TAN({column:0})"
  let atan = column -> s"ATAN({column:0})"
  let pow = exponent column -> s"POW({column:0}, {exponent:0})"
  let round = n_digits column -> s"ROUND({column:0}, {n_digits:0})"
}

# Other functions
let as = `type` column -> s"CAST({column:0} AS {type:0})"

# Text functions
module text {
  # Clickhouse: https://clickhouse.com/docs/en/sql-reference/functions/string-functions
  # DuckDB: https://duckdb.org/docs/sql/functions/char
  # MySQL: https://dev.mysql.com/doc/refman/8.0/en/string-functions.html
  # Postgres: https://www.postgresql.org/docs/7.4/functions-string.html
  # SQLite: https://www.sqlite.org/lang_corefunc.html
  # MSSQL: https://learn.microsoft.com/en-us/sql/t-sql/functions/string-functions-transact-sql
  # BigQuery: https://cloud.google.com/bigquery/docs/reference/standard-sql/string_functions
  # Snowflake: https://docs.snowflake.com/en/sql-reference/functions-string
  let lower = column -> s"LOWER({column:0})"
  let upper = column -> s"UPPER({column:0})"
  let ltrim = column -> s"LTRIM({column:0})"
  let rtrim = column -> s"RTRIM({column:0})"
  let trim = column -> s"TRIM({column:0})"
  let length = column -> s"CHAR_LENGTH({column:0})"
  let extract = offset length column -> s"SUBSTRING({column:0}, {offset:0}, {length:0})"
  let replace = pattern replacement column -> s"REPLACE({column:0}, {pattern:0}, {replacement:0})"
  let starts_with = prefix column -> s"{column:0} LIKE CONCAT({prefix:0}, '%')"
  let contains = substr column -> s"{column:0} LIKE CONCAT('%', {substr:0}, '%')"
  let ends_with = suffix column -> s"{column:0} LIKE CONCAT('%', {suffix:0})"
}

# Source-reading functions, primarily for DuckDB
let read_parquet = source -> s"read_parquet({source:0})"
let read_csv = source -> s"read_csv({source:0})"

@{binding_strength=11}
let mul = l r -> null

@{binding_strength=100}
let div_i = l r -> s"FLOOR(ABS({l:11} / {r:12})) * SIGN({l:0}) * SIGN({r:0})"

# We have a simple float division by default, but it can be overridden by
# dialects.
# Note that this uses `12` for the RHS binding strength, which is one more than
# binding strength of division. That's because we don't use associativity here,
# and so we need to make sure that the RHS is parenthesized when the binding
# strengths are equal; e.g. in `a / (b / c)`.
@{binding_strength=11}
let div_f = l r -> s"{l} / {r:12}"

@{binding_strength=11}
let mod = l r -> s"{l} % {r:12}"

@{binding_strength=10}
let add = l r -> null

@{binding_strength=10}
let sub = l r -> null

@{binding_strength=6}
let eq = l r -> null

@{binding_strength=6}
let ne = l r -> null

@{binding_strength=6}
let gt = l r -> null

@{binding_strength=6}
let lt = l r -> null

@{binding_strength=6}
let gte = l r -> null

@{binding_strength=6}
let lte = l r -> null

@{binding_strength=3}
let and = l r -> null

@{binding_strength=2}
let or = l r -> null

let coalesce = l r -> s"COALESCE({l:0}, {r:0})"

let regex_search = text pattern -> s"REGEXP({text:0}, {pattern:0})"

@{binding_strength=13}
let neg = l -> s"-{l}"

@{binding_strength=4}
let not = l -> s"NOT {l}"

module ansi {
  @{binding_strength=11}
  let div_f = l r -> s"({l} * 1.0 / {r:12})"
}

module bigquery {
  @{binding_strength=11}
  let div_f = l r -> s"({l} * 1.0 / {r:12})"

  # Mathematical functions
  module math {
    # BigQuery: https://cloud.google.com/bigquery/docs/reference/standard-sql/mathematical_functions
    let degrees = column -> s"({column:0} * 180 / PI())"
    let radians = column -> s"({column:0} * PI() / 180)"
  }

  let regex_search = text pattern -> s"REGEXP_CONTAINS({text:0}, {pattern:0})"
}

module clickhouse {
  # https://clickhouse.com/docs/en/sql-reference/functions/arithmetic-functions#divide
  @{binding_strength=11}
  let div_f = l r -> s"({l} / {r:12})"

  @{binding_strength=11}
  let div_i = l r -> s"({l} DIV {r:12})"

  # Date functions
  module date {
    # https://clickhouse.com/docs/en/sql-reference/functions/date-time-functions
    let to_text = format column -> s"formatDateTimeInJodaSyntax({column:0}, {format:0})"
  }

  let regex_search = text pattern -> s"match({text:0}, {pattern:0})"

  let read_csv = source -> s"file({source:0}, 'CSV')"

  let read_parquet = source -> s"file({source:0}, 'Parquet')"
}

module duckdb {
  @{binding_strength=11}
  let div_f = l r -> s"({l} / {r:12})"

  @{binding_strength=11}
  let div_i = l r -> s"TRUNC({l:11} / {r:12})"

  # Text functions
  module text {
    # DuckDB: https://duckdb.org/docs/sql/functions/char
    let length = column -> s"LENGTH({column:0})"
  }

  # Date functions
  module date {
    # https://duckdb.org/docs/sql/functions/dateformat
    let to_text = format column -> s"strftime({column:0}, {format:0})"
  }

  let regex_search = text pattern -> s"REGEXP_MATCHES({text:0}, {pattern:0})"

  let read_csv = source -> s"read_csv_auto({source:0})"
}

module mssql {
  @{binding_strength=11}
  let div_f = l r -> s"({l} * 1.0 / {r:12})"

  # Mathematical functions
  module math {
    # https://learn.microsoft.com/en-us/sql/t-sql/functions/mathematical-functions-transact-sql
    let ceil = column -> s"CEILING({column:0})"
    let ln = column -> s"LOG({column:0})"
    let pow = exponent column -> s"POWER({column:0}, {exponent:0})"
  }

  # Text functions
  module text {
    # https://learn.microsoft.com/en-us/sql/t-sql/functions/string-functions-transact-sql
    let length = column -> s"LEN({column:0})"
  }

  # Date functions
  module date {
    # https://learn.microsoft.com/en-us/sql/t-sql/functions/format-transact-sql
    let to_text = format column -> s"FORMAT({column:0}, {format:0})"
  }

  let regex_search = text pattern -> null
}

module mysql {
  @{binding_strength=11}
  let div_f = l r -> s"({l} / {r:12})"

  @{binding_strength=11}
  let div_i = l r -> s"({l} DIV {r:12})"

  @{binding_strength=100}
  let mod = l r -> s"ROUND(MOD({l:0}, {r:0}))"

  # Date functions
  module date {
    # https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html
    let to_text = format column -> s"DATE_FORMAT({column:0}, {format:0})"
  }

  # 'c' for case-sensitive
  let regex_search = text pattern -> s"REGEXP_LIKE({text:0}, {pattern:0}, 'c')"
}

module postgres {
  @{binding_strength=11}
  let div_f = l r -> s"({l} * 1.0 / {r:12})"

  @{binding_strength=100}
  let div_i = l r -> s"TRUNC({l:11} / {r:12})"

  # Mathematical functions
  module math {
    # Postgres: https://www.postgresql.org/docs/7.4/functions-math.html
    @{binding_strength=100}
    let round = n_digits column -> s"ROUND(({column:0})::numeric, {n_digits:0})"
  }

  # Text functions
  module text {
    # Postgres: https://www.postgresql.org/docs/7.4/functions-string.html
    let extract = offset length column -> s"SUBSTR({column:0}, {offset:0}, {length:0})"
  }

  # Date functions
  module date {
    # https://www.postgresql.org/docs/current/functions-formatting.html
    let to_text = format column -> s"TO_CHAR({column:0}, {format:0})"
  }

  @{binding_strength=9}
  let regex_search = text pattern -> s"{text} ~ {pattern}"
}

module glaredb {
  @{binding_strength=11}
  let div_f = l r -> s"({l} * 1.0 / {r:12})"

  @{binding_strength=100}
  let div_i = l r -> s"TRUNC({l:11} / {r:12})"

  # Mathematical functions
  module math {
    # GlareDB: https://glaredb.com/docs/sql-reference/functions/math-functions
    @{binding_strength=100}
    let round = n_digits column -> s"ROUND(({column:0})::numeric, {n_digits:0})"
  }

  @{binding_strength=9}
  let regex_search = text pattern -> s"{text} ~ {pattern}"

  let read_csv = source -> s"csv_scan({source:0})"

  let read_parquet = source -> s"parquet_scan({source:0})"
}

module sqlite {
  @{window_frame=true, coalesce="TRUE", binding_strength=6}
  let all = column -> s"MIN({column:0}) > 0"

  @{window_frame=true, coalesce="FALSE", binding_strength=6}
  let any = column -> s"MAX({column:0}) > 0"

  @{window_frame=true, coalesce="''"}
  let concat_array = column -> s"GROUP_CONCAT({column:0}, '')"

  @{binding_strength=11}
  let div_f = l r -> s"({l} * 1.0 / {r:12})"

  @{binding_strength=100}
  let div_i = l r -> s"ROUND(ABS({l:11} / {r:12}) - 0.5) * SIGN({l:0}) * SIGN({r:0})"

  # Text functions
  module text {
    # SQLite: https://www.sqlite.org/lang_corefunc.html
    let length = column -> s"LENGTH({column:0})"
    let starts_with = prefix column -> s"{column:0} LIKE {prefix:0} || '%'"
    let contains = substr column -> s"{column:0} LIKE '%' || {substr:0} || '%'"
    let ends_with = suffix column -> s"{column:0} LIKE '%' || {suffix:0}"
  }

  @{binding_strength=9}
  let regex_search = text pattern -> s"{text} REGEXP {pattern}"
}

module snowflake {
  # https://docs.snowflake.com/en/sql-reference/operators-arithmetic#division
  @{binding_strength=11}
  let div_f = l r -> s"({l} / {r:12})"
}
