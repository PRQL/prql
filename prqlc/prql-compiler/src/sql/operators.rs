use std::collections::HashMap;
use std::iter::zip;
use std::path::PathBuf;

use anyhow::Result;
use itertools::Itertools;
use once_cell::sync::Lazy;

use super::gen_expr::{translate_operand, ExprOrSource, SourceExpr};
use super::{Context, Dialect};

use crate::ir::{decl, pl, rq};
use crate::semantic;
use crate::utils::Pluck;
use crate::{Error, WithErrorInfo};

static STD: Lazy<decl::Module> = Lazy::new(load_std_sql);

fn load_std_sql() -> decl::Module {
    let std_lib = crate::SourceTree::new([(
        PathBuf::from("std.prql"),
        include_str!("./std.sql.prql").to_string(),
    )]);
    let ast = crate::parser::parse(&std_lib).unwrap();

    let options = semantic::ResolverOptions {};

    let context = semantic::resolve(ast, options).unwrap();
    context.module
}

pub(super) fn translate_operator_expr(expr: rq::Expr, ctx: &mut Context) -> Result<ExprOrSource> {
    let (name, args) = expr.kind.into_operator().unwrap();

    let source = translate_operator(name, args, ctx).with_span(expr.span)?;

    Ok(ExprOrSource::Source(source))
}

pub(super) fn translate_operator(
    name: String,
    args: Vec<rq::Expr>,
    ctx: &mut Context,
) -> Result<SourceExpr> {
    let (func_def, binding_strength, window_frame, coalesce) =
        find_operator_impl(&name, ctx.dialect_enum).unwrap();
    let parent_binding_strength = binding_strength.unwrap_or(100);

    let params = func_def
        .named_params
        .iter()
        .chain(func_def.params.iter())
        .map(|x| x.name.split('.').last().unwrap_or(x.name.as_str()));

    let args: HashMap<&str, _> = zip(params, args).collect();

    // body can only be an s-string
    let body = match &func_def.body.kind {
        pl::ExprKind::Literal(pl::Literal::Null) => {
            return Err(Error::new_simple(format!(
                "operator {} is not supported for dialect {}",
                name, ctx.dialect_enum
            ))
            .into())
        }
        pl::ExprKind::SString(items) => items,
        _ => panic!("Bad RQ operator implementation. Expected s-string or null"),
    };

    let mut text = String::new();

    for item in body {
        match item {
            pl::InterpolateItem::Expr { expr, format } => {
                // s-string exprs can only contain idents
                let ident = expr.kind.as_ident();
                let ident = ident.as_ref().unwrap();

                // lookup args
                let arg = args.get(ident.name.as_str()).unwrap().clone();

                // binding strength
                let required_strength = format
                    .as_ref()
                    .and_then(|f| f.parse::<i32>().ok())
                    .unwrap_or(parent_binding_strength);

                // translate args
                let arg = translate_operand(
                    arg,
                    false,
                    required_strength,
                    super::gen_expr::Associativity::Both,
                    ctx,
                )?;

                text += &arg.into_source();
            }
            pl::InterpolateItem::String(s) => {
                text += s;
            }
        }
    }

    let mut binding_strength = parent_binding_strength;

    if !ctx.query.window_function {
        if let Some(default) = coalesce {
            text = format!("COALESCE({text}, {default})");
            binding_strength = 100;
        }
    }

    Ok(SourceExpr {
        text,
        binding_strength,
        window_frame,
    })
}

fn find_operator_impl(
    operator_name: &str,
    dialect: Dialect,
) -> Option<(&pl::Func, Option<i32>, bool, Option<String>)> {
    let operator_name = operator_name.strip_prefix("std.").unwrap();
    let operator_ident = pl::Ident::from_path(
        operator_name
            .split('.')
            .map(String::from)
            .collect::<Vec<_>>(),
    );

    let dialect_module = STD.get(&pl::Ident::from_name(dialect.to_string()));

    let mut func_def = None;

    if let Some(dialect_module) = dialect_module {
        let module = dialect_module.kind.as_module().unwrap();
        func_def = module.get(&operator_ident);
    }

    if func_def.is_none() {
        func_def = STD.get(&operator_ident);
    }

    let decl = func_def?;

    let func_def = decl.kind.as_expr().unwrap();
    let func_def = func_def.kind.as_func().unwrap();

    let mut annotation = decl
        .clone()
        .annotations
        .into_iter()
        .exactly_one()
        .ok()
        .and_then(|x| x.tuple_items().ok())
        .unwrap_or_default();

    let binding_strength = pluck_annotation(&mut annotation, "binding_strength")
        .and_then(|literal| literal.into_integer().ok())
        .map(|int| int as i32);

    let window_frame = pluck_annotation(&mut annotation, "window_frame")
        .and_then(|literal| literal.into_boolean().ok())
        .unwrap_or_default();

    let coalesce =
        pluck_annotation(&mut annotation, "coalesce").and_then(|val| val.into_string().ok());

    Some((func_def.as_ref(), binding_strength, window_frame, coalesce))
}

fn pluck_annotation(
    annotation: &mut Vec<(String, pl::ExprKind)>,
    name: &str,
) -> Option<pl::Literal> {
    annotation
        .pluck(|(n, val)| if n == name { Ok(val) } else { Err((n, val)) })
        .into_iter()
        .next()
        .and_then(|val| val.into_literal().ok())
}
