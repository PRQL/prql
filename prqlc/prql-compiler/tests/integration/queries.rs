#![cfg(not(target_family = "wasm"))]
use std::path::Path;
use std::{env, fs};

use insta::{assert_snapshot, with_settings};

use prql_compiler::sql::Dialect;
use prql_compiler::{Options, Target};
use test_each_file::test_each_path;

mod compile {
    use super::*;

    // If this is giving compilation errors saying `expected identifier, found keyword`,
    // rename the filenames in queries to something that's not a keyword in Rust.
    test_each_path! { in "./prqlc/prql-compiler/tests/integration/queries" => run }

    fn run(prql_path: &Path) {
        let test_name = prql_path.file_stem().unwrap().to_str().unwrap();
        let prql = fs::read_to_string(prql_path).unwrap();
        if prql.contains("generic:skip") {
            return;
        }

        let target = Target::Sql(Some(Dialect::Generic));
        let options = Options::default().no_signature().with_target(target);

        let sql = prql_compiler::compile(&prql, &options).unwrap();

        with_settings!({ input_file => prql_path }, {
            assert_snapshot!(test_name, &sql, &prql)
        });
    }
}

mod fmt {
    use super::*;

    test_each_path! { in "./prqlc/prql-compiler/tests/integration/queries" => run }

    fn run(prql_path: &Path) {
        let test_name = prql_path.file_stem().unwrap().to_str().unwrap();
        let prql = fs::read_to_string(prql_path).unwrap();

        let pl = prql_compiler::prql_to_pl(&prql).unwrap();
        let formatted = prql_compiler::pl_to_prql(pl).unwrap();

        with_settings!({ input_file => prql_path }, {
            assert_snapshot!(test_name, &formatted, &prql)
        });
    }
}

#[cfg(any(feature = "test-dbs", feature = "test-dbs-external"))]
mod results {
    use super::*;

    use std::{ops::DerefMut, sync::Mutex};

    use anyhow::Context;
    use once_cell::sync::Lazy;
    use prql_compiler::sql::SupportLevel;

    use crate::dbs::{ConnectionCfg, DbConnection, DbProtocol};

    static CONNECTIONS: Lazy<Mutex<Vec<DbConnection>>> = Lazy::new(init_connections);

    fn init_connections() -> Mutex<Vec<DbConnection>> {
        let configs = [
            ConnectionCfg {
                dialect: Dialect::SQLite,
                data_file_root: "tests/integration/data/chinook".to_string(),

                protocol: DbProtocol::SQLite,
            },
            ConnectionCfg {
                dialect: Dialect::DuckDb,
                data_file_root: "tests/integration/data/chinook".to_string(),

                protocol: DbProtocol::DuckDb,
            },
            ConnectionCfg {
                dialect: Dialect::Postgres,
                data_file_root: "/tmp/chinook".to_string(),

                protocol: DbProtocol::Postgres {
                    url: "host=localhost user=root password=root dbname=dummy".to_string(),
                },
            },
            ConnectionCfg {
                dialect: Dialect::MySql,
                data_file_root: "/tmp/chinook".to_string(),

                protocol: DbProtocol::MySql {
                    url: "mysql://root:root@localhost:3306/dummy".to_string(),
                },
            },
            ConnectionCfg {
                dialect: Dialect::ClickHouse,
                data_file_root: "chinook".to_string(),

                protocol: DbProtocol::MySql {
                    url: "mysql://default:@localhost:9004/dummy".to_string(),
                },
            },
            ConnectionCfg {
                dialect: Dialect::GlareDb,
                data_file_root: "/tmp/chinook".to_string(),

                protocol: DbProtocol::Postgres {
                    url: "host=localhost user=glaredb dbname=glaredb port=6543".to_string(),
                },
            },
            ConnectionCfg {
                dialect: Dialect::MsSql,
                data_file_root: "/tmp/chinook".to_string(),

                protocol: DbProtocol::MsSql,
            },
        ];

        let mut connections = Vec::new();
        for cfg in configs {
            if !matches!(
                cfg.dialect.support_level(),
                SupportLevel::Supported | SupportLevel::Unsupported
            ) {
                continue;
            }

            // The filtering is not a great design, since it doesn't proactively
            // check that we can get connections; but it's a compromise given we
            // implement the external_dbs feature using this.
            let Some(mut connection) = DbConnection::new(cfg) else {
                continue;
            };

            connection.setup();

            connections.push(connection);
        }
        Mutex::new(connections)
    }

    test_each_path! { in "./prqlc/prql-compiler/tests/integration/queries" => run }

    fn run(prql_path: &Path) {
        let test_name = prql_path.file_stem().unwrap().to_str().unwrap();
        let prql = fs::read_to_string(prql_path).unwrap();

        let data_file_root_keyword = "data_file_root";
        let is_contains_data_root = prql.contains(data_file_root_keyword);

        // for each of the connections
        let mut results = Vec::new();
        for con in CONNECTIONS.lock().unwrap().deref_mut() {
            if !con.should_run_query(&prql) {
                continue;
            }

            let mut prql = prql.clone();
            if is_contains_data_root {
                prql = prql.replace(data_file_root_keyword, &con.cfg.data_file_root);
            }

            let dialect = con.cfg.dialect;

            println!("Executing {test_name} for {dialect}");
            let rows = con
                .run_query(&prql)
                .context(format!("Executing {test_name} for {dialect}"))
                .unwrap();

            // convert into ad-hoc CSV
            let result = rows
                .iter()
                .map(|r| r.join(","))
                .collect::<Vec<_>>()
                .join("\n");

            results.push((dialect, result));
        }

        // insta::allow_duplicates!, but with reporting of which two cases are not matching.
        let (left_dialect, left_text) = results.swap_remove(0);
        for (right_dialect, right_text) in results {
            similar_asserts::assert_eq!(
                left_text,
                right_text,
                "{} {} {}",
                test_name,
                left_dialect,
                right_dialect
            );
        }

        with_settings!({ input_file => prql_path }, {
            assert_snapshot!(test_name, left_text, &prql)
        })
    }
}
