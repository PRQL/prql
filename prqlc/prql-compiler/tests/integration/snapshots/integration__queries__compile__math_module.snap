---
source: prqlc/prql-compiler/tests/integration/queries.rs
expression: "# mssql:test\n# sqlite:skip (see https://github.com/rusqlite/rusqlite/issues/1211)\nfrom invoices\ntake 5\nselect {\n    total_original = total | math.round 2,\n    total_x = math.pi - total | math.round 2 | math.abs,\n    total_floor = math.floor total,\n    total_ceil = math.ceil total,\n    total_log10 = math.log10 total | math.round 3,\n    total_log2 = math.log 2 total | math.round 3,\n    total_sqrt = math.sqrt total | math.round 3,\n    total_ln = math.ln total | math.exp | math.round 2,\n    total_cos = math.cos total | math.acos | math.round 2,\n    total_sin = math.sin total | math.asin | math.round 2,\n    total_tan = math.tan total | math.atan | math.round 2,\n    total_deg = total | math.degrees | math.radians | math.round 2,\n    total_square = total | math.pow 2| math.round 2,\n}\n"
input_file: prqlc/prql-compiler/tests/integration/queries/math_module.prql
---
SELECT
  ROUND(total, 2) AS total_original,
  ABS(ROUND(PI() - total, 2)) AS total_x,
  FLOOR(total) AS total_floor,
  CEIL(total) AS total_ceil,
  ROUND(LOG10(total), 3) AS total_log10,
  ROUND(LOG10(total) / LOG10(2), 3) AS total_log2,
  ROUND(SQRT(total), 3) AS total_sqrt,
  ROUND(EXP(LN(total)), 2) AS total_ln,
  ROUND(ACOS(COS(total)), 2) AS total_cos,
  ROUND(ASIN(SIN(total)), 2) AS total_sin,
  ROUND(ATAN(TAN(total)), 2) AS total_tan,
  ROUND(RADIANS(DEGREES(total)), 2) AS total_deg,
  ROUND(POW(total, 2), 2) AS total_square
FROM
  invoices
LIMIT
  5

